imagine/main.go

```diff
@@ -0,0 +1,263 @@
+)
+
+var (
+	client *pilosa.Client
+)
+
+// UsersConfig describes the configuration for the `users` index.
+type UsersConfig struct {
+	Columns int64 `help:"number of columns in users table"`
+	Numbers int   `help:"number of rows of favorite-numbers"`
+}
+
+// EventsConfig describes the configuration for the `events` index.
+type EventsConfig struct {
+	Columns int64 `help:"number of columns in events table"`
+}
+
+// Config describes the overall configuration of the tool.
+type Config struct {
+	Host   string `help:"host name for Pilosa server"`
+	Port   int    `help:"host port for Pilosa server"`
+	Create bool   `help:"create indexes if needed"`
+	Delete bool   `help:"delete existing indexes with same prefix"`
+	Prefix string `help:"prefix to use on index names"`
+	Users  UsersConfig
+	Events EventsConfig
+}
+
+// Run does validation on the configuration data. Used by
+// commandeer.
+func (c *Config) Run() error {
+	// no error-checking if nothing to check errors on
+	if c == nil {
+		return nil
+	}
+	if int(uint16(c.Port)) != c.Port {
+		return fmt.Errorf("port %d out of range", c.Port)
+	}
+	if len(c.Prefix) == 0 {
+		return errors.New("index name prefix must not be empty")
+	}
+
+	// If you don't specify, default to creating but not deleting
+	// existing indexes.
+	if !c.Create && !c.Delete {
+		c.Create = true
+	}
+	return nil
+}
+
+func main() {
+	// Conf defines the default/initial values for config, which
+	// can be overridden by command line options.
+	conf := &Config{
+		Host: "localhost",
+		Port: 10101,
+		Users: UsersConfig{
+			Columns: 1000,
+			Numbers: 50,
+		},
+		Events: EventsConfig{},
+		Prefix: "imaginary",
+	}
+
+	err := commandeer.Run(conf)
+	if err != nil {
+		log.Fatalf("err: %s", err)
+	}
+
+	uri, err := pilosa.NewURIFromHostPort(conf.Host, uint16(conf.Port))
+	if err != nil {
+		log.Fatalf("could not create Pilosa URI: %s", err)
+	}
+
+	client, err = pilosa.NewClient(uri)
+	if err != nil {
+		log.Fatalf("could not create Pilosa client: %s", err)
+	}
+
+	prefix := conf.Prefix + "-"
+	for suffix, indexBuilder := range indexBuilders {
+		indexBuilder.name = prefix + suffix
+	}
+
+	if conf.Delete {
+		err = deleteIndexes(client, conf)
+		if err != nil {
+			log.Fatal(err.Error())
+		}
+	}
+
+	if conf.Create {
+		err = createIndexes(client, conf)
+		if err != nil {
+			log.Fatal(err.Error())
+		}
+	}
+}
+
+type indexData struct {
+	name      string
+	builder   indexBuilder
+	populator indexBuilder
+}
+type indexBuilder func(client *pilosa.Client, conf *Config, index *pilosa.Index) error
+
+var indexBuilders = map[string]*indexData{
+	"users":        &indexData{builder: makeUsers, populator: populateUsers},
+	"events":       &indexData{builder: makeEvents},
+	"supplemental": &indexData{builder: makeSupplemental},
+}
+
+func deleteIndexes(client *pilosa.Client, conf *Config) error {
+	schema, err := client.Schema()
+	if err != nil {
+		return fmt.Errorf("could not retrieve schema: %s", err)
```

Comment:
We have been standardizing on `errors.Wrap(err, "retrieving schema")` in our other projects. I just wanted to note that here in case it's useful to continue that pattern.

Reply:
that's probably reasonable, i was avoiding adding a dependency, but don't really object to pkg/errors.

Reply:
it's practically stdlib ;)

Reply:
that's honestly my biggest objection to pkg/errors, i think the name is misleading. it looks a lot more official than it is.